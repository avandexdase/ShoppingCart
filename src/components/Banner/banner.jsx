import React, { memo, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { useSwipeable } from 'react-swipeable';

function Banner({ banner }) {
  const [carouselId, setCarouselId] = useState(1);
  const [paused, setPaused] = useState(false);
  function plusSlides(n) {
    showSlides(carouselId + n);
  }
  useEffect(() => {
    const interval = setInterval(() => {
      if (!paused) showSlides(carouselId + 1);
    }, 2000);
    return () => {
      if (interval) {
        clearInterval(interval);
      }
    };
  });
  const handlers = useSwipeable({
    onSwipedLeft: () => showSlides(carouselId + 1),
    onSwipedRight: () => showSlides(carouselId - 1),
  });

  function currentSlide(n) {
    showSlides(n);
  }

  function showSlides(n) {
    let slideIndex = 1;
    if (n > banner.length) {
      slideIndex = 1;
    } else if (n < 1) {
      slideIndex = banner.length;
    } else slideIndex = n;
    setCarouselId(slideIndex);
  }
  return banner.length > 0 ? (
    <>
      <div className="slideshow-container">
        {banner.map((each) => (
          <div
            className={
              carouselId == each.order ? 'swiper__item_active' : 'swiper__item'
            }
            key={each.order}
            onMouseEnter={() => setPaused(true)}
            onMouseLeave={() => setPaused(false)}
            {...handlers}
          >
            <img
              src={each.bannerImageUrl}
              className="swiper__img"
              alt={each.bannerImageAlt}
            />
          </div>
        ))}

        <span className="prev" onClick={() => plusSlides(-1)}>
          PREV
        </span>
        <span className="next" onClick={() => plusSlides(1)}>
          NEXT
        </span>
      </div>

      <div className="dot_container">
        {banner.map((each) => (
          <span
            className={each.order === carouselId ? 'dot_active' : 'dot'}
            key={each.order}
            onClick={() => currentSlide(each.order)}
          ></span>
        ))}
      </div>
    </>
  ) : null;
}

Banner.propTypes = {
  banner: PropTypes.array,
};

export default memo(Banner);
