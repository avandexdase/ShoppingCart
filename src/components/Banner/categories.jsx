import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { useNavigate } from 'react-router-dom';
function Categories({ data, index }) {
  const navigate = useNavigate();
  const handleSubmit = (each) => {
    navigate('/products', { state: { id: each.id } });
  };
  return (
    <div key={data.id} className="home_categories">
      <img
        className="home_categories__img"
        src={data.imageUrl}
        alt={data.name}
        data-testid="homecategoryImg"
      />
      {/* <div className="col-start-4 col-span-1 sm:hidden"></div> */}
      <div className="home_categories__Item">
        <h2 className="home_categories__Name" data-testid="homecategoryName">
          {data.name}
        </h2>
        <p className="home_categories__desc" data-testid="homecategoryDesc">
          {data.description}
        </p>
        <button
          className="home_categories__explorebtn"
          onClick={() => handleSubmit(data)}
        >
          {`explore ${data.key}`}
        </button>
      </div>
    </div>
  );
}

Categories.propTypes = {
  data: PropTypes.exact({
    name: PropTypes.string,
    key: PropTypes.string,
    description: PropTypes.string,
    enabled: PropTypes.bool,
    order: PropTypes.number,
    imageUrl: PropTypes.string,
    id: PropTypes.string,
  }),
  index: PropTypes.number,
};

export default memo(Categories);
