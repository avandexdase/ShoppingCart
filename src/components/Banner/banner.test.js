import React from 'react';
import {
  screen,
  render,
  waitFor,
  cleanup,
  waitForElement,
} from '@testing-library/react';
import Banner from '../../components/Banner/banner';
import '@testing-library/jest-dom';
import axios from 'axios';
jest.mock('axios');

const data = [
  {
    bannerImageUrl: '/static/images/offers/offer1.jpg',
    bannerImageAlt: 'Independence Day Deal - 25% off on shampoo',
    isActive: true,
    order: 1,
    id: '5b6c38156cb7d770b7010ccc',
  },
  {
    bannerImageUrl: '/static/images/offers/offer2.jpg',
    bannerImageAlt: 'Independence Day Deal - Rs120 off on surf',
    isActive: true,
    order: 2,
    id: '5b6c38336cb7d770b7010ccd',
  },
];
describe('<Banner />', () => {
  afterEach(cleanup);
  it('renders banner image', () => {
    const { getByAltText } = render(<Banner banner={data} />);
    const image = getByAltText('Independence Day Deal - 25% off on shampoo');
    expect(image).toHaveAttribute('src', '/static/images/offers/offer1.jpg');
  });
  it('test prev-next btn', () => {
    render(<Banner banner={data} />);
    expect(screen.getByTestId('prevBtn')).toBeInTheDocument();
    expect(screen.getByTestId('nextBtn')).toBeInTheDocument();
  });
  it('renders home page correctly', async () => {
    axios.get.mockResolvedValue({
      data,
    });
    const { getByTestId, asFragment } = render(<Banner banner={data} />);
    const listNode = await waitFor(() => getByTestId('banner'));
    expect(listNode.children).toHaveLength(4);
    expect(asFragment()).toMatchSnapshot();
  });
});
