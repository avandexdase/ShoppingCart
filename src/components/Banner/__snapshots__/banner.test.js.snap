// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Banner /> renders home page correctly 1`] = `
<DocumentFragment>
  <div
    class="slideshow-container"
    data-testid="banner"
  >
    <div
      class="swiper__item_active"
    >
      <img
        alt="Independence Day Deal - 25% off on shampoo"
        class="swiper__img"
        src="/static/images/offers/offer1.jpg"
      />
    </div>
    <div
      class="swiper__item"
    >
      <img
        alt="Independence Day Deal - Rs120 off on surf"
        class="swiper__img"
        src="/static/images/offers/offer2.jpg"
      />
    </div>
    <span
      class="prev"
      data-testid="prevBtn"
    >
      PREV
    </span>
    <span
      class="next"
      data-testid="nextBtn"
    >
      NEXT
    </span>
  </div>
  <div
    class="dot_container"
  >
    <span
      class="dot_active"
    />
    <span
      class="dot"
    />
  </div>
</DocumentFragment>
`;
